File: CP/LeetCode/Python/144.py
# 144. Binary Tree Preorder Traversal

from __future__ import annotations


class TreeNode:
    def __init__(
        self, val: int = 0, left: TreeNode | None = None, right: TreeNode | None = None
    ):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def preorderTraversal(self, root: TreeNode) -> list[int]:
        # Time Complexity of recursive solution is O(n) and Space Complexity is O(n)

        # if not root:
        #     return []
        # return [root.val] + self.preorderTraversal(root.left) + self.preorderTraversal(root.right)

        # ---

        # Time Complexity of iterative solution is O(n) and Space Complexity is O(n)

        res = []
        stack = [root] if root else []

        while stack:
            _root = stack.pop()
            res.append(_root.val)
            if _root.right:
                stack.append(_root.right)
            if _root.left:
                stack.append(_root.left)

        return res


if __name__ == "__main__":
    new_root = TreeNode(1, None, TreeNode(2, TreeNode(3)))
    sol = Solution()

    print(sol.preorderTraversal(new_root))

File: CP/LeetCode/Java/144.java
// 144. Binary Tree Preorder Traversal

import java.util.ArrayList;
import java.util.List;

class Solution {
    static class TreeNode {
        // CSOFF: VisibilityModifier
        int val;
        TreeNode left;
        TreeNode right;
        // CSON: VisibilityModifier

        TreeNode(final int x) {
            this.val = x;
        }
    }

    public List<Integer> preorderTraversal(final TreeNode root) {
        List<Integer> res = new ArrayList<>();
        preorder(root, res);
        return res;
    }

    private void preorder(final TreeNode root, final List<Integer> res) {
        if (root == null) {
            return;
        }
        res.add(root.val);
        preorder(root.left, res);
        preorder(root.right, res);
    }

    public static void main(final String[] args) {
        Solution sol = new Solution();
        TreeNode root = new TreeNode(1);
        root.right = new TreeNode(2);
        root.right.left = new TreeNode(3);
        System.out.println(sol.preorderTraversal(root));
    }
}
